modules/powers/Hecate.class.php

```diff
@@ -13,10 +13,250 @@ public function __construct($game, $playerId)
+       || $log['action'] == 'moveToken')
+      { 
+        $args = json_decode($log['action_arg'], true);
+        $space = $args['to'];
+      }
+      elseif ($log['action'] == 'removal')
+      {
+        $space = $this->game->board->getPiece($log['piece_id']);
+      }
+      else
+        return null;
+       
+      if (max(array_map(function($s) use ($space) {
+                      return ($space['x'] == $s['x'] && $space['y'] == $s['y']) ;
+                      } , $myWorkers )))
+         return $space;
+      else
+        return null;
+  
+  }
+  
+  
+  // check if the turn was legal based on Hecate power, and cancel the last actions if necessary
+  // parameter: for Maenads
+  public function endOpponentTurn($testOnly = false) 
+  {
+    $myWorkers = $this->getPlacedWorkers();
+    
+    $logs = $this->game->log->logsForCancelTurn();
+    
+    $space = null;
+    
+    foreach (array_reverse($logs) as $log) {
+        $space = $this->getAffectedSpaceFromLog($log, $myWorkers);
+        if ($space == null)
+          continue;
+        // cancel end of turn
+        
+        if ($testOnly)
+          return true;
+        $moveIds = $this->game->log->cancelTurn($log['log_id']);    
+        break;
+    }
+    
+    if ($testOnly)
+      return false;
+    
+    if ($space == null)
+    {
+      // treat Medusa: kill secret workers only after we know the turn is legal
+      $powers = $this->game->playerManager->getPlayer($this->game->getActivePlayerId())->getPowers();
+      foreach ($powers as $power)
+      {
+        if ($power->getId() != MEDUSA)
+          continue;
+        $argKill = ['workers' => []];
+        $power->argPlayerBuild($argKill, true); // get killable secret workers
+        $power->endPlayerTurn($argKill); // kill them
+      }
+      return;
+    }
+    
+    
+    
+    // display current board state
+    $playerIds = $this->game->playerManager->getPlayerIds();
+    foreach ($playerIds as $playerId) {
+      $this->game->notifyPlayer($playerId, 'cancel', '' , [
+        'placedPieces' => $this->game->board->getPlacedPieces($playerId),
+          'moveIds' => $moveIds,
+        ]);
+     }
+    
+    // display the problematic secret worker
+    $args = [
+      'i18n' => ['player_name'],
+      'player_name' => $this->game->getActivePlayerName(),
+      'power_name' => $this->getName(),
+      'coords' => $this->game->board->getMsgCoords($space)
+    ];
+    
+    $dummy = $this->getDummyWorker(); // use a dummy one to avoid duplicates
+    $dummy['x'] = $space['x'];
+    $dummy['y'] = $space['y'];
+    $dummy['z'] = $this->game->board->countBlocksAt($space);
+    $this->game->notifyAllPlayers('revealPiece', '', ['piece' => $dummy]);
+    
+    // explain what is happening
+    $this->game->notifyAllPlayers('message', '${power_name}: due to a secret worker in ${coords}, an action of ${player_name} is illegal which cancels the rest of the turn', $args);
```

Comment:
I think the display is improved if we show the message first (with the 'cancel' notification), so the user can read what is happening while the secret worker animation runs.